### DISCUSSION TITLE
Lua filter to wrap each paragraph in a custom-style "Paragraph" div, and apply "New paragraph" div for paragraphs after a blockquote or a heading or title (beginning) for Word template export

### DISCUSSION THREAD

**Discussion by ElDuderino321** [jgm/pandoc#10124] [Q&A] [answered]
I'm new to writing Lua filters (have zero programming experience), but I managed to write a filter to turn Blockquotes into a custom-style div for a word template I need to use:

`function BlockQuote(elem)
  return pandoc.Div (elem.content, {["custom-style"] = "Displayed quotation"})
end`

This was easy. But now, I want to wrap every paragraph. The way I understand it, I would need to somehow distinguish "Para" elements in the AST that are part of the top-level, but exclude any Para that is nested inside a deeper element. Also (in a second script probably), I would need to test for the conditions of a "New paragraph," and then change the newly made "Paragraph" divs into that.

But how can I do that? I am still learning Lua, and I think the solution (at least for the first part) has something to do with "walking the AST?"

Any tips would be greatly appreciated!

**Comment by alerque** [comment, marked as answer]
Use the `Blocks` filter and iterate over the top level blocks yourself instead of letting Pandoc throw `Para` blocks to your filter willy-nilly. Change them to what you want them to be in place, something like this:

```lua
local function Blocks (blocks)
   for i, block in ipairs(blocks) do
      if block.t == "Para" then
         blocks[i] = pandoc.Div(block.content, {...})
      end
   end
   return blocks
end
```

**Reply by ElDuderino321** [reply]
Awesome, thanks! That's the most important part of my question!! I had a faint inkling it might have to do with Blocks. But got a lot of learning to do.

How can I test for what comes before a paragraph though?

**Reply by alerque** [reply]
Use the iteration counter to grab the previous index out of the list, e.g. to find the first paragraph after hea#ings:

```lua
if i >= 2 and block.t == "Para" and blocks[i-1].t == "Heading" then
```

**Reply by ElDuderino321** [reply]
Thank you again! This will go a long way to help me understand how to come up with this myself, too.

**Reply by ElDuderino321** [reply]
> Use the `Blocks` filter and iterate over the top level blocks yourself instead of letting Pandoc throw `Para` blocks to your filter willy-nilly. Change them to what you want them to be in place, something like this:
> 
> ```lua
> local function Blocks (blocks)
>    for i, block in ipairs(blocks) do
>       if block.t == "Para" then
>          blocks[i] = pandoc.Div(block.content, {...})
>       end
>    end
>    return blocks
> end
> ```

I actually only now got around to trying this out, and in the output, every word is now its own paragraph (albeit in the correct style). Looking at the resulting AST, each div contains Plain elements, which contains the Str and Space elements. Is that related to the problem?

(I also had to remove the "local" for the filter to function. Otherwise, it did not nothing. Is that a pandoc thing?)
